### PR TITLE
feat(base): implement `Chained`

### PR DESCRIPTION
**Pull request by TomTaehoonKim**
# Description

This PR implements chain over iterators

### CODE REVIEW

**Comment by chokobole** [comment]
Please update the commit title feat(base): implement chain to  feat(base): implement `Chained`

File: tachyon/base/containers/adapters.h [hunk 1]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+          end_second_(end_second),
+          is_first_(true),
+          value_(*it_) {}
+
+    ItType it_;
+    const ItType end_first_;
+    const ItType begin_second_;
+    const ItType end_second_;
+    bool is_first_;
+    value_type value_;
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:279]: Please remove this

File: tachyon/base/containers/adapters.h [hunk 2]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+
+   private:
+    Iterator(ItType it, ItType end_first, ItType begin_second,
+             ItType end_second)
+        : it_(it),
+          end_first_(end_first),
+          begin_second_(begin_second),
+          end_second_(end_second),
+          is_first_(true),
+          value_(*it_) {}
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:272]: Please remove this

File: tachyon/base/containers/adapters.h [hunk 3]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+        if (it_ == end_first_) {
+          it_ = begin_second_;
+          is_first_ = false;
+        }
+      } else {
+        if (it_ != end_second_) {
+          ++it_;
+        }
+      }
+      value_ = *it_;
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:250]: Please remove this

File: tachyon/base/containers/adapters.h [hunk 4]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+      return it;
+    }
+
+    pointer operator->() { return &value_; }
+
+    reference operator*() { return value_; }
+
+   private:
+    Iterator(ItType it, ItType end_first, ItType begin_second,
+             ItType end_second)
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:266]:
```suggestion
    Iterator(ItType it, ItType end_first, ItType begin_second,
             ItType end_second, bool is_first)
```

File: tachyon/base/containers/adapters.h [hunk 5]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+
+    using difference_type = std::ptrdiff_t;
+    using value_type = iter_value_t<ItType>;
+    using pointer = value_type*;
+    using reference = value_type&;
+    using iterator_category = std::forward_iterator_tag;
+
+    static Iterator begin(T& first, T& second) {
+      return Iterator(std::begin(first), std::end(first), std::begin(second),
+                      std::end(second));
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:225]:
```suggestion
      return Iterator(std::begin(first), std::end(first), std::begin(second),
                      std::end(second), true);
```

File: tachyon/base/containers/adapters.h [hunk 6]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+    }
+    static Iterator end(T& first, T& second) {
+      return Iterator(std::end(second), std::end(first), std::begin(second),
+                      std::end(second));
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:229]:
```suggestion
      return Iterator(std::begin(first), std::end(first), std::begin(second),
                      std::end(second), false);
```

**Review by chokobole — CHANGES_REQUESTED** [review]

File: tachyon/base/containers/adapters.h [hunk 7]
```diff
@@ -206,6 +206,93 @@ class ZippedAdapter {
...
+
+    Iterator operator++(int) {
+      Iterator it(*this);
+      ++(*this);
+      return it;
+    }
+
+    pointer operator->() { return &value_; }
+
+    reference operator*() { return value_; }
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:262]: Use `it_` directly

File: tachyon/base/containers/adapters.h (hunk 6 above)
Comment (TomTaehoonKim) [reply tachyon/base/containers/adapters.h:229]: Isn't this correct since it's supposed to be the end?

Comment (chokobole) [reply tachyon/base/containers/adapters.h:229]: When you iterate from `begin()` to `end()`, then `is_first` is changed to `false`. So I think `flase` is the right value for the `end()`.

Comment (TomTaehoonKim) [reply tachyon/base/containers/adapters.h:229]: @chokobole Oh I meant the suggestion you've made to change to std::begin(first).

File: tachyon/base/containers/adapters.h [hunk 8]
```diff
@@ -206,6 +206,92 @@ class ZippedAdapter {
...
+      return *this;
+    }
+
+    Iterator operator++(int) {
+      Iterator it(*this);
+      ++(*this);
+      return it;
+    }
+
+    pointer operator->() { return it_; }
```
Comment (chokobole) [inline tachyon/base/containers/adapters.h:261]: Please refer to other implementations to overcome `-Werror,-Wignored-reference-qualifiers`

Plus,  the type `pointer` isn't matched to `iterator`, to match it,

```suggestion
    pointer operator->() { return &*it_; }
```

File: tachyon/base/containers/adapters.h (hunk 6 above)
Comment (chokobole) [reply tachyon/base/containers/adapters.h]: Oh okay got it!

**Comment by TomTaehoonKim** [comment]
Since the implementation is not perfect in terms of multiple chains, and we don't need it right away, I'm converting this PR to draft.

**Pull request closed without merging**